include/pmacc/lockstep/ForEach.hpp

```diff
@@ -120,26 +128,31 @@ namespace pmacc
              */
             template<
                 typename T_Functor,
-                // check if functor is invocable
-                typename = Void<decltype(alpaka::core::declval<T_Functor>()(alpaka::core::declval<Idx const>()))>,
+                // check if functor is invokable
+                typename = IsCallableWithIndex<T_Functor>,
                 std::
                     enable_if_t<std::is_void<InvokeResult_t<T_Functor, Idx const>>::value && domainSize != 1, int> = 0>
             HDINLINE void operator()(T_Functor&& functor) const
             {
-                for(uint32_t i = 0u; i < numCollIter; ++i)
+                static constexpr uint32_t numFullLoops = domainSize / (simdSize * numWorkers);
+                for(uint32_t i = 0u; numFullLoops != 0u && i < numFullLoops; ++i)
                 {
                     uint32_t const beginWorker = i * simdSize;
                     uint32_t const beginIdx = beginWorker * numWorkers + simdSize * this->getWorkerIdx();
-                    if(outerLoopCondition || !innerLoopCondition || beginIdx < domainSize)
+                    for(uint32_t s = 0u; s < simdSize; ++s)
+                        functor(Idx(beginIdx + s, beginWorker + s));
+                }
+
+                // process left over indices if the domainSize is not a multiple of 'simdSie * numWorkers'
+                static constexpr bool hasRemainder = (domainSize % (simdSize * numWorkers)) != 0u;
+                static constexpr uint32_t leftOverIndices = domainSize - (numFullLoops * numWorkers * simdSize);
+                for(uint32_t s = 0u; hasRemainder && s < simdSize; ++s)
+                    if(hasRemainder && this->getWorkerIdx() * simdSize + s < leftOverIndices)
```

Comment:
If we are here, `hasRemainder` is true. Seems no need to check in the `if` statement.

Reply:
If this is executed we do not know if we need to calculate `simdSize` steps or less.
If e.g. simdSize is `2` it could be the first index is within the domain the second is not.
`hasRemainder` is required to hide comparison warning because all values are known at compile time. I should add a comment above.

Reply:
From  the readability point of view, I think the best would be to just compute a proper ending index and have a simple loop from 0 to that. I assume you didn't go for it, since it's additional registers?

Reply:
Unfortionally this loop condition is only allowed to have compile-time values therefore we have the runtime condition as `if` condition in the body.
This allows the compiler to perform range checks and map access at compile time.